### PR TITLE
ENH: stats.monte_carlo_test: add array API support

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Towards gh-20544

#### What does this implement/fix?
This PR adds array-API support for `scipy.stats.monte_carlo_test`.

#### Additional information
Example (15~50x speedup locally, depending on dtype):

```python3
import numpy as np
import cupy as cp
from scipy import stats

rng = np.random.default_rng()
data_np = rng.standard_normal(size=100)
data_cp64 = cp.asarray(data_np, dtype=cp.float64)
data_cp32 = cp.asarray(data_np, dtype=cp.float32)

ref = stats.ttest_1samp(data_np, 0.)

def get_statistic(xp):
    def statistic(x, axis=0):
        m = xp.mean(x, axis=axis)
        # this is just the user-defined statistic;
        # it doesn't need to be array-API compliant
        v = xp.var(x, axis=axis, ddof=1)
        n = x.shape[axis]
        return m / (v / n) ** 0.5
    return statistic

# NumPy 64-bit
rng_np = np.random.default_rng()
res1 = stats.monte_carlo_test(data_np, rng_np.standard_normal, get_statistic(np), n_resamples=99999)
# 123 ms ± 1.02 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

# CuPy 64-bit
rng_cp = cp.random.default_rng()
res2 = stats.monte_carlo_test(data_cp64, rng_cp.standard_normal, get_statistic(cp), n_resamples=99999)
# 8.27 ms ± 767 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

# CuPy 32-bit
def rvs(size):
    return rng_cp.standard_normal(size=size, dtype=cp.float32)
res3 = stats.monte_carlo_test(data_cp32, rvs, get_statistic(cp), n_resamples=99999)
# 2.51 ms ± 33.4 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

print(ref.pvalue, res1.pvalue, res2.pvalue, res3.pvalue)
# 0.7506365948009475 0.754 0.75104 0.74768
```

### CODE REVIEW

File: scipy/stats/_resampling.py
Comment (mdhaber) [inline scipy/stats/_resampling.py:961]: array-api-strict doesn't like adding booleans or, if if `cmps` is integer dtype, performing `__truediv__`. So convert directly to the result dtype. I think p-values are still coming out as `float64`, though... I'll look into it.

File: scipy/stats/tests/test_resampling.py
Comment (mdhaber) [inline scipy/stats/tests/test_resampling.py:1]: Again, didn't want to add to the mess of imports.

File: scipy/stats/tests/test_resampling.py
Comment (mdhaber) [inline scipy/stats/tests/test_resampling.py:749]: This works for CPU, but I wanted to test GPU, too.

File: scipy/stats/tests/test_resampling.py
Comment (mdhaber) [inline scipy/stats/tests/test_resampling.py:810]: New input validation tests for the new error messages.

File: scipy/stats/tests/test_resampling.py
Comment (mdhaber) [inline scipy/stats/tests/test_resampling.py:843]: I got lazy and didn't create a separate `xp_test`. We need array-api-compat version because the statistic uses `var` with `correction`. Hopefully that's OK.

**Review by mdhaber — COMMENTED** [review]
Some self-review to aid the reviewer.

File: scipy/stats/_resampling.py
Comment (mdhaber) [reply scipy/stats/_resampling.py:961]: Added dtype-maintenance check to `test_axis_dtype`.

File: scipy/_lib/_array_api.py
Comment (j-bowhay) [inline scipy/_lib/_array_api.py:361]: I notice we are accumulating quite a lot of these, I think in terms of the bigger picture (ie not within the scope of this PR) I think it would be quite nice to have a way to document somewhere what we have and their status in the array API (are they intended to be temporary or not)

Comment (mdhaber) [reply scipy/_lib/_array_api.py:361]: The ones I added here are in the 2023 standard; they just aren't in the libraries or our `array-api-compat` yet.

Comment (rgommers) [reply scipy/_lib/_array_api.py:361]: This should materialize in `array-api-compat` within weeks I expect. Fine to commit to our `_array_api.py` now; should be temporary.

File: scipy/_lib/_array_api.py
Comment (rgommers) [reply scipy/_lib/_array_api.py:374]: Same, `clip` is in the 2023 version of the standard and should be available in `array-api-compat` soon.

File: scipy/stats/_resampling.py
Comment (rgommers) [reply scipy/stats/_resampling.py:687]: This is indeed a bit ugly. try-except for input validation is usually better written some other way. Doesn't something like this work:
```
if not isinstance(data, collections.Sequence):
    raise TypeError(f"`data` should be a sequence if `rvs` is a tuple of callables, but found `data` is: {type(data)}")
```

Also it seems to me that `TypeError` is the correct error type here.

File: scipy/_lib/_array_api.py
Comment (rgommers) [reply scipy/_lib/_array_api.py:384]: Shouldn't it be called `xp_moveaxis` then? Should also be available in `array-api-compat` soon.

**Comment by rgommers** [comment]
> Example (15~50x speedup locally, depending on dtype):

Very nice! This diff should be pretty clean once `array-api-compat` supports v2023.12 of the standard.

File: scipy/stats/_resampling.py
Comment (rgommers) [reply scipy/stats/_resampling.py:687]: Also note that `len(data)` may work on some array types but not all. It also tends to not be super readable, I always have to remind myself whether it returns `.shape[0]` or something else. So best never to use it on arrays.

Comment (mdhaber) [reply scipy/stats/_resampling.py:687]: Yes, this PR changes another instance of `len` to `shape[0]` for that reason.

The trouble here is that when this function added support for multiple sample statistics, the user was *supposed* to pass a "sequence" here:
> An array or sequence of arrays of observations.

But the function did not strictly require that this argument be a `Sequence` of arrays; it only requires that it have length equal to that of `rvs` and that it be iterable. 

NumPy arrays are not instances of `collections.abc.Sequence`, so users who use a single array as a collection of arrays along axis 0 would start getting this error.

One option is to leave this and raise a deprecation warning afterwards if the argument is not a `Sequence`, then we could just do a `Sequence` check at the end of the deprecation cycle.

File: scipy/_lib/_array_api.py
Comment (mdhaber) [reply scipy/_lib/_array_api.py:384]: If it supported an `destination` argument, yes, but I only wrote it to move an axis to the end.

**Comment by mdhaber** [comment]
Thanks for taking a look @rgommers; responses are above. Another thing you might have thoughts about is the idea of making the default tolerance of `xp_assert_close` dtype-dependent.
https://github.com/scipy/scipy/pull/20597#pullrequestreview-2033062044
 The intend is to save us from constantly forcing torch arrays to be `float64` to meet tolerances and/or setting tolerances for different dtypes manually.

File: scipy/stats/_resampling.py
Comment (rgommers) [reply scipy/stats/_resampling.py:687]: +1 for deprecation that. Using an N+1-dim array as "sequence of N-dim arrays" is very ugly and I assume wasn't intended to ever work.

**Comment by rgommers** [comment]
> Towards https://github.com/scipy/scipy/issues/20544

If `stats` functions are going to be done one by one over many PRs rather than in one go, would it make sense to have a separate tracker for `stats` to point to? Having an overview of the coverage in `stats` will be useful (and help contributors pick functions to work on), and it's a bit much to put that overview in the scipy-wide tracker probably.

**Comment by mdhaber** [comment]
> would it make sense to have a separate tracker for stats to point to?

? You referred to the separate stats tracker.

File: scipy/stats/_resampling.py
Comment (mdhaber) [reply scipy/stats/_resampling.py:687]: Ok. I might do that in a follow-up up PR if this one is otherwise ready or very close.

**Comment by rgommers** [comment]
> ? You referred to the separate stats tracker.

oops, sorry for the noise. temporary blindness there

**Comment by mdhaber** [comment]
@rgommers I know you'll be out for a bit - shall I ask others to review/merge this in the meantime, or did you want to continue with it?

**Comment by rgommers** [comment]
Thanks for asking. Please don't feel like you have to wait for me.

File: scipy/stats/tests/test_resampling.py
Comment (j-bowhay) [reply scipy/stats/tests/test_resampling.py:750]: Has this been done?

File: scipy/stats/tests/test_resampling.py
Comment (j-bowhay) [inline scipy/stats/tests/test_resampling.py:818]: Can this not be done with the standard testing infrastructure?

File: scipy/stats/tests/test_resampling.py
Comment (j-bowhay) [inline scipy/stats/tests/test_resampling.py:933]: Can you add a comment explaining why the tests below haven't been converted? I guess they are waiting on other things?

File: scipy/stats/tests/test_resampling.py
Comment (mdhaber) [reply scipy/stats/tests/test_resampling.py:750]: Yup, I was testing with CuPy when I wrote this.

File: scipy/stats/tests/test_resampling.py
Comment (mdhaber) [reply scipy/stats/tests/test_resampling.py:818]: I'm not sure I follow what the link is adding. `raises` can check for partial matches, sure, but it requires that the exception be raised. This exception only occurs when introspection is not possible (e.g. PyTorch). The comments notes:

> *If* this raises a value error, make sure it has the intended message

File: scipy/stats/tests/test_resampling.py
Comment (mdhaber) [inline scipy/stats/tests/test_resampling.py:932]:
```suggestion
    # Tests below involve statistics that are not yet array-API compatible.
    # They can be converted when the statistics are converted.
    @pytest.mark.slow
```

File: scipy/stats/tests/test_resampling.py
Comment (mdhaber) [reply scipy/stats/tests/test_resampling.py:933]: Sure; suggested above.

**Review by mdhaber — COMMENTED** [review]
Replied to comments. Thanks @j-bowhay!

File: scipy/stats/tests/test_resampling.py
Comment (lucascolley) [inline scipy/stats/tests/test_resampling.py:844]: can we add a comment for why this is needed here too? (thanks for adding in the other places)

**Review by lucascolley — COMMENTED** [review]
this all looks idiomatic to me, at a glance at least 👍

**Review by j-bowhay — COMMENTED** [review]
Think this should be good to go once the merge conflicts and Lucas's comment is resolved

**Review by j-bowhay — APPROVED** [review]

**Comment by mdhaber** [comment]
Thanks, all! I think I'll do `power` next, because it is quite similar and doesn't require an RNG.

**Pull request merged**
